tests: add full, automated, top-level test runner for hardware

dpgeorge opened the pull request:
### Summary

With the aim to speed up testing for a release, I put together a script to automatically run tests on devices running MicroPython.  We have a lot of different tests these days, with different ways to run them.  The idea with this top-level test runner is:
- detect all serial devices connected
- detect what platform is running on each device and what its capabilities are (eg WLAN, BLE, ability to import .mpy files)
- run as many tests as possible given the connected devices (eg if there are two WLAN boards connected, then run the network multitests against them, among other things)

Basically all you need to do is connect boards to your PC via USB and run:
```
$ cd tests
$ ./hwtest.py
```
Then it does the rest.  It takes between 10 and 15 minutes per board, so if you have many connected it can take a few hours.  You can also restrict to given devices by specifying them on the command line, eg:
```
$ ./hwtest.py a0 a1 u0
```
(Probably the name of the script `hwtest.py` can change to something better.)

### Testing

This was used to test 16 distinct boards for the recent v1.24.0 release.

### Trade-offs and Alternatives

This is a bit of a hacky script at the moment but can be improved over time.

Currently there are no boards that pass all tests (!!) and so over time we need to fix things and improve the tests so more of them can pass.

dpgeorge committed:
tests/run-tests.py: Don't print info for tests that pass.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-multitests.py: Suppress diff.

TODO: make it a CLI option

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Add a full, automatic test runner.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
PR #16111 will help simplify this test runner a bit.

PR #15909 is included in this PR.

hmaerki commented:
Octoprobe is ready so far to run these tests fully automated, including build of all firmware variants and to be triggered from github workflows.

We have to agree on triggers (github action on PR, github manual action). I will start the discussion when I am back from vacation.

dpgeorge committed:
tests/run-tests.py: Capture KeyboardInterrupt for nicer exit.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Make sure unittest is installed.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Add command to build all natmods.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/micropython/extreme_exc.py: Skip on ESP32 with PSRAM.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Improve CLI options.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests: Increase buffer to 249 bytes.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Add option to not print list of skipped tests.

To reduce output clutter.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Just print returncode on subprocess fail.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Change args to selection then devices.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Run hardware tests in native mode if selected.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Support targets without a filesystem.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Exit with code 2 on error.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Run hardware tests after native tests.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Increase sleep and fix list append bug.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Add multi_wlan tests.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Use -j8 for building natmods.

dpgeorge committed:
tests/hwtest.py: Remove armv6 and armv7em.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Fix target check.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Skip Octoprobe infrastructure.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Allow selecting natmods to build.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/multi_extmod: Update pins for I2CTarget test using octoprobe.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests: Improve test support for renesas-ra boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware/machine_i2c_target.py: Improve pyboard support.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Decrease MP_STREAM_GET_BUFFER_SIZE to 32-7.

If it's too big PYBLITEV10 will run out of memory when running
`extmod/deflate_decompress.py`.

Actually, 32-7 is the default... so this ioctl could be removed.  Really
need to benchmark it to see if it makes any difference having it large.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Add argparse and support a reference board.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Support any number of return values from target_info.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod/machine_uart_tx.py: Add tunings for zephyr boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring/UM_FEATHERS2.py: Add target_wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring/ZEPHYR_FRDM_K64F.py: Add target_wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Update to use --test-instance.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Use new serial_test.py.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
wip

dpgeorge committed:
tests/run-multitests.py: Ignore DHCPS messages.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Update for new serial_test.py.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/ports/stm32_hardware/spi_dma_align.py: Convert to target_wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware: Add Pin test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests//micropython/schedule_sleep.py: Fail on nrf, samd.

Fixed on tests-improvements-for-octoprobe.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
wip

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring: Add more boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring: Add target wiring for NUCLEO_G0B1RE, NUCLEO_L152RE.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring/PYBx.py: Update target wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/ports/stm32/uart.py: Skip on NUCLEO_U575.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware/machine_adc.py: Add machine.ADC test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring: Update target wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-multitests.py: Save results to .exp/.out files.

There's already a --results-dir argument, the directory is already made,
and a _results.json is already put there.  So adding .exp/.out makes sense.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring: Update target wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring/esp32.py: Update esp32 target wiring.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: WIP adding esp8266 faster REPL baudrate.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Improve.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/hwtest.py: Improve.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware/machine_pwm.py: Use sleep_ms so no need for float.

TODO: can we remove this esp32-specific code?

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware/machine_pin_irq.py: Add Pin.irq test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests: Improve machine.Pin tests.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring/NUCLEO_U575ZI_Q.py: WIP

dpgeorge committed:
tune machine_pin_irq test

dpgeorge committed:
tests/target_wiring/samd.py: Add pin_loopback_pins.

pin.irq doesn't yet work, seems it files rising and falling regardless of
config.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Add --trace-output option.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/target_wiring/rp2.py: Add wiring for pins tests.

Signed-off-by: Damien George <damien@micropython.org>

Gadgetoid commented:
On macOS it seems to hiccup on `/dev/cu.Bluetooth-Incoming-Port`:

```
 ./hwtest.py   
WLAN_SSID and/or WLAN_PASS not provided
Selected tests: serial, bytecode, hardware, via-mpy, native, wlan, ble

================================================================
TARGETS TO BE TESTED
detect_octoprobe(/dev/cu.Bluetooth-Incoming-Port)
b''
Traceback (most recent call last):
  File "micropython/micropython/tests/./hwtest.py", line 447, in <module>
    main()
    ~~~~^^
  File "micropython/micropython/tests/./hwtest.py", line 334, in main
    targets = list_targets(selected_devices, [ref_target.device] if ref_target else [])
  File "micropython/micropython/tests/./hwtest.py", line 124, in list_targets
    if detect_octoprobe(p.device):
       ~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "micropython/micropython/tests/./hwtest.py", line 96, in detect_octoprobe
    t.enter_raw_repl()
    ~~~~~~~~~~~~~~~~^^
  File "micropython/micropython/tests/../tools/mpremote/mpremote/transport_serial.py", line 179, in enter_raw_repl
    raise TransportError("could not enter raw repl")
mpremote.transport.TransportError: could not enter raw repl
```

Should probably catch these exceptions and continue, like this but less grotty:

```diff
diff --git a/lib/nrfx b/lib/nrfx
index 11f57e578..7a4c9d946 160000
--- a/lib/nrfx
+++ b/lib/nrfx
@@ -1 +1 @@
-Subproject commit 11f57e578c7feea13f21c79ea0efab2630ac68c7
+Subproject commit 7a4c9d946cf1801771fc180acdbf7b878f270093
diff --git a/tests/hwtest.py b/tests/hwtest.py
index c53b65a01..52d714ae6 100755
--- a/tests/hwtest.py
+++ b/tests/hwtest.py
@@ -12,7 +12,7 @@ import time
 import serial_test
 
 sys.path.append("../tools/mpremote")
-from mpremote.transport_serial import SerialTransport
+from mpremote.transport_serial import SerialTransport, TransportError
 import mpremote.commands
 import mpremote.mip
 
@@ -91,6 +91,7 @@ def map_device_name(d):
 
 
 def detect_octoprobe(device):
+    print(f"detect_octoprobe({device})")
     t = SerialTransport(device)
     t.enter_raw_repl()
     t.exec("import os")
@@ -106,9 +107,13 @@ def list_targets(args, exclude_devices):
         for device in args:
             for p in valid_serial_ports:
                 if device == p.device:
-                    if detect_octoprobe(device):
-                        print(f"{device} Octoprobe infrastructure device")
-                    targets.append(Target(device, p.serial_number))
+                    try:
+                        if detect_octoprobe(device):
+                            print(f"{device} Octoprobe infrastructure device")
+                        targets.append(Target(device, p.serial_number))
+                    except TransportError:
+                        print(f"{device} could not connect!")
+                        continue
                     break
             else:
                 print(f"{device} not found")
@@ -120,9 +125,13 @@ def list_targets(args, exclude_devices):
                 continue
             if p.device in exclude_devices:
                 continue
-            if detect_octoprobe(p.device):
-                # If no devices specified, skip Octoprobe infrastructure.
-                print(f"{p.device} Octoprobe infrastructure device, skipping")
+            try:
+                if detect_octoprobe(p.device):
+                    # If no devices specified, skip Octoprobe infrastructure.
+                    print(f"{p.device} Octoprobe infrastructure device, skipping")
+                    continue
+            except TransportError:
+                print(f"{p.device} could not connect!")
                 continue
             targets.append(Target(p.device, p.serial_number))
     return targets
```
